source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2397 @@
+solicit a ``readPreference`` argument to target nodes other than the primary?
+To date, no spec needs this behavior and this change could easily be deferred to
+a future minor version of the test format.
+
+
+Representing options in operation.arguments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Existing spec tests are inconsistent in how they represent optional
+parameters for operations. In some cases, optional parameters appear directly
+in `operation.arguments`_ alongside required parameters (e.g. CRUD). In other
+cases, they are nested under an ``options`` key (e.g. Sessions).
+
+Most specs *are* consistent about how API methods are defined in spec documents.
+The CRUD spec was on the first documents to define API methods and can be
+credited with introducing the ``options`` parameter (e.g.
+``options: Optional<UpdateOptions>``); however, this was likely done for
+readability of the spec (avoiding very long method declarations and/or embedding
+``Optional`` syntax therein). Additionally, documenting options in a separate
+type (e.g. UpdateOptions) allows the declarations to be shared and reused across
+multiple methods.
+
+That said, this syntax is in no way prescriptive for language implementations.
+While some drivers to model options as a struct/object, others solicit them
+alongside required parameters (e.g. Python's keyword/named arguments).
+
+Should this spec require that optional parameters be nested under an ``options``
+key or solicit them directly in `operation.arguments`_? Both are technically
+possible, since test runners handle both forms today.
+
+With regard to CRUD, this issue dates back to
+`SPEC-1158 <https://jira.mongodb.org/browse/SPEC-1158>`__. Therein, Jeremy
+originally argued in favor of nesting under an ``options`` key because it was
+most consistent the text in existing spec documents.
+
+Note: the CRUD spec WriteModels (e.g. UpdateOneModel) for ``bulkWrite`` to not
+use ``options`` keys in either the spec document or test files. As such, the
+resolution of this question would not impact how WriteModels are expressed in
+test files (see: `bulkWrite`_).
+
+Note: this question does not pertain to TransactionOptions in SessionOptions,
+which is always nested under ``defaultTransactionOptions``. This is a special
+case where all drivers represent TransactionOptions as a separate struct/object,
+even if they do not do so for ``startTransaction``.
+
+
+Change Stream Tests
+-------------------
+
+
+Error Assertions
+~~~~~~~~~~~~~~~~
+
+`Iterating the Change Stream <../change-streams/tests#iterating-the-change-stream>`__
+in the change stream spec test documentation states:
+
+  If the test expects an error and one was not thrown by either creating the
+  change stream or executing the test's operations, iterating the change stream
+  once allows for an error to be thrown by a getMore command.
+
+It's not clear if this is necessary because the existing change stream spec test
```

Comment:
I think async drivers might follow a pattern where the initial aggregate query is only run when the change stream is first iterated. Motor, at least, definitely follows this pattern - the call to `watch()` does no I/O and creates an object that lazily creates the change stream when results are sought by the application. I would be interested to know whether this is also the case for Node @mbroadst ?
Pinpointing when a change stream is created will be essential in writing reliable change stream tests, to avoid, for instance, missing events when change streams are initialized lazily in situations like this:
```
cs = collection.watch()
collection.insert_many([...])
cs.next()
```

We could consider using a special operator like `createChangeStream` instead of a vanilla `watch()` to start a change stream. This operator would stipulate that runners establish the change stream cursor server-side and store the client-side object in the entity map. I'd favor this approach over requiring drivers to conditionally perform extra operations when `watch()` is run because there is no way to figure out simply by reading the YAML test that the `watch` operation does anything special. Readers would need to consult the spec to see that watch() is special-cased which is not intuitive. Using a special command would make this much more intuitive for the test reader to discover.

Reply:
`createChangeStream` seems like a good compromise.

Alternatively, if `watch` _always_ returns a ChangeStream, we could also introduce a new custom operation on the change stream entity (e.g. `ensureCursorIsEstablished`), which would be a no-op for drivers that _do_ establish a cursor when `watch` is called; however, we'd then have to insert that into every test file (more verbose) and it'd still require special handling so I think the custom operation you suggested is preferable. We can also explain that test files SHOULD NOT use `watch` for that reason.

Reply:
Added `createChangeStream` operation for client, database, and collection and advised test files not to use `watch` directly.